hw/drivers/sensors/lis2dw12/src/lis2dw12.c

```diff
@@ -1863,60 +1915,92 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
 int lis2dw12_run_self_test(struct sensor_itf *itf, int *result)
 {
     int rc;
-    int16_t base[3], pos[3];
+    /*configure min and max values for reading 5 samples, and accounting for
+     * both negative and positive offset */
+    int min = LIS2DW12_ST_MIN*5*2;
+    int max = LIS2DW12_ST_MAX*5*2;
+
+    int16_t data[3], diff[3] = {0,0,0};
     int i;
-    int16_t change;
-    uint8_t fs;
+    uint8_t prev_config[6];
+    /* set config per datasheet, with positive self test mode enabled. */
+    uint8_t st_config[] = {0x44, 0x08, 0x40, 0x00, 0x00, 0x10};
 
-    rc = lis2dw12_get_fs(itf, &fs);
+    rc = lis2dw12_readlen(itf, LIS2DW12_REG_CTRL_REG1, prev_config, 6);
     if (rc) {
         return rc;
     }
-
-    /* ensure self test mode is disabled */
-    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_DISABLE);
+    rc = lis2dw12_writelen(itf, LIS2DW12_REG_CTRL_REG2, &st_config[1], 5);
+    rc = lis2dw12_writelen(itf, LIS2DW12_REG_CTRL_REG1, st_config, 1);
     if (rc) {
         return rc;
     }
 
-    os_time_delay(OS_TICKS_PER_SEC / 10);
-
-    /* take base reading */
-    rc = lis2dw12_get_data(itf, fs, &(base[0]), &(base[1]), &(base[2]));
+    /* go into self test mode 1 */
+    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_MODE1);
     if (rc) {
         return rc;
     }
 
-    /* set self test mode to positive self test */
-    rc = lis2dw12_set_self_test(itf, LIS2DW12_ST_MODE_MODE1);
+    /* wait 100ms */
+    os_time_delay(OS_TICKS_PER_SEC / 100);
+    rc = lis2dw12_get_data(itf, 2, &(data[0]), &(data[1]), &(data[2]));
```

Comment:
Should not be using hardcoded FS value, but rather the FS_4G define, at all the other lis2dw12_get_data calls too